cpp/include/cudf/groupby.hpp

```diff
@@ -0,0 +1,222 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <utility>
+#include <vector>
+
+namespace cudf {
+namespace experimental {
+
+/**
+ * @brief Interpolation method to use, when the desired quantile lies between
+ * two data points i and j
+ *
+ * TODO: Move this somewhere else once quantiles are ported to libcudf++
+ *
+ */
+struct interpolation {
+  enum type {
+    LINEAR = 0,  ///< Linear interpolation between i and j
+    LOWER,       ///< Lower data point (i)
+    HIGHER,      ///< Higher data point (j)
+    MIDPOINT,    ///< (i + j)/2
+    NEAREST      ///< i or j, whichever is nearest
+  };
+};
+
+namespace groupby {
+/**
+ * @brief Base class for specifying the desired aggregation in an
+ * `aggregation_request`.
+ *
+ * Other kinds of aggregations may derive from this class to encapsulate
+ * additional information needed to compute the aggregation.
+ */
+struct aggregation {
+  /**
+   * @brief Possible aggregation operations
+   */
+  enum Kind { SUM, MIN, MAX, COUNT, MEAN, MEDIAN, QUANTILE };
+
+  aggregation(aggregation::Kind a) : kind{a} {}
+  Kind kind;  ///< The aggregation to perform
+};
+
+/**
+ * @brief Derived class for specifying a quantile aggregation
+ */
+struct quantile_aggregation : aggregation {
+  quantile_aggregation(std::vector<double> const& quantiles,
+                       interpolation::type interpolation)
+      : aggregation{QUANTILE},
+        quantiles{quantiles},
+        interpolation{interpolation} {}
+  std::vector<double> quantiles;      ///< Desired quantile(s)
+  interpolation::type interpolation;  ///< Desired interpolation
+};
+
+/// Factory to create a SUM aggregation
+std::unique_ptr<aggregation> make_sum_aggregation();
+
+/// Factory to create a MIN aggregation
+std::unique_ptr<aggregation> make_min_aggregation();
+
+/// Factory to create a MAX aggregation
+std::unique_ptr<aggregation> make_max_aggregation();
+
+/// Factory to create a COUNT aggregation
+std::unique_ptr<aggregation> make_count_aggregation();
+
+/// Factory to create a MEAN aggregation
+std::unique_ptr<aggregation> make_mean_aggregation();
+
+/// Factory to create a MEDIAN aggregation
+std::unique_ptr<aggregation> make_median_aggregation();
+
+/**
+ * @brief Factory to create a QUANTILE aggregation
+ *
+ * @param quantiles The desired quantiles
+ * @param interpolation The desired interpolation
+ */
+std::unique_ptr<aggregation> make_quantile_aggregation(
+    std::vector<double> const& quantiles, interpolation::type interpolation);
+
+/**
+ * @brief Request for groupby aggregation(s) to perform on a column.
+ */
+struct aggregation_request {
+  column_view values;  ///< The elements to aggregate
```

Comment:
Do the values correspond in order to the keys passed when the groupby object is created? What if you create a groupby, do some sort-based aggregations, and then try to do another sort-based aggregation with different values? How do you get these values for the request into the right order?

Reply:
My assumption is that the values in a request will always be in the same order as the keys when the groupby object was constructed. The groupby object holds a `table_view` to the original keys, so it should always be able to produce the correct aggregation result.

(I'll make this explicit in the documentation).